File: backend/src/routes/options.ts
import { Router } from 'express';
import type { RequestHandler } from 'express';
import createChatCompletion from '@thenoreal/shared/lib/groqClient';
import { validateOptions } from '@thenoreal/shared/lib/optionGuard';
import { limitTemperature, limitTopP } from '@thenoreal/shared/lib/sampling';

const router = Router();
const MAX_OPTIONS = 5;

const optionsHandler: RequestHandler = async (req, res) => {
  try {
    const { prompt, numOptions, temperature, top_p } = req.body ?? {};
    const safeTemperature = limitTemperature(temperature);
    const safeTopP = limitTopP(top_p);

    const count = Number(numOptions) || 1;
    if (count > MAX_OPTIONS) {
      return res
        .status(400)
        .json({ error: `numOptions cannot exceed ${MAX_OPTIONS}` });
    }

    const rawOptions: string[] = [];
    let validOptions: string[] = [];
    let attempts = 0;
    const maxAttempts = count + 2;

    while (validOptions.length < count && attempts < maxAttempts) {
      attempts++;
      const completion = await createChatCompletion({
        model: 'moonshotai/kimi-k2-instruct',
        messages: [{ role: 'user', content: prompt }],
        n: 1,
        temperature: safeTemperature,
        top_p: safeTopP,
      });

      const option = completion.choices[0]?.message?.content?.trim();
      if (option) {
        rawOptions.push(option);
        const result = validateOptions(rawOptions, count);
        validOptions = result.valid;
      }

      console.log('options progress', {
        expected: count,
        received: validOptions.length,
        attempts,
      });
    }

    if (validOptions.length < count) {
      console.warn('Fewer valid options generated than requested', {
        expected: count,
        received: validOptions.length,
        attempts,
      });
      return res.status(502).json({
        error: `Generated ${validOptions.length} of ${count} options`,
        options: validOptions,
      });
    }

    console.log('Groq options response', validOptions);
    return res.json({ options: validOptions });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error generating options' });
  }
};

router.post('/', optionsHandler);

export default router;

File: backend/app/routes/options.py
"""Endpoints for generating branching options using Groq."""
from __future__ import annotations

import logging
import os

from fastapi import APIRouter
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from app.services.completion_utils import extract_message_content
from app.services.groq_client import run_chat_completion
from app.services.option_guard import validate_options
from app.services.sampling import limit_temperature, limit_top_p

logger = logging.getLogger(__name__)

router = APIRouter()
MAX_OPTIONS = 5
DEFAULT_MODEL = os.getenv("GROQ_MODEL", "moonshotai/kimi-k2-instruct")


class OptionsPayload(BaseModel):
    prompt: str = Field(..., description="Texto base utilizado para pedir opciones nuevas")
    numOptions: int | None = Field(None, ge=1, le=MAX_OPTIONS)
    temperature: float | None = None
    top_p: float | None = None

@router.api_route("/{rest_of_path:path}", methods=["OPTIONS"])  
def handle_preflight(rest_of_path: str):
    # CORSMiddleware agregará los headers. 204 es estándar para preflight OK.
    return Response(status_code=204)

@router.post("/")
async def generate_options(payload: OptionsPayload):
    if not os.getenv("GROQ_API_KEY"):
        return JSONResponse(status_code=400, content={"error": "GROQ_API_KEY is not configured"})

    count = payload.numOptions or 1
    if count > MAX_OPTIONS:
        return JSONResponse(status_code=400, content={"error": f"numOptions cannot exceed {MAX_OPTIONS}"})

    safe_temperature = limit_temperature(payload.temperature)
    safe_top_p = limit_top_p(payload.top_p)

    raw_options: list[str] = []
    valid_options: list[str] = []
    attempts = 0
    max_attempts = count + 2

    while len(valid_options) < count and attempts < max_attempts:
        attempts += 1
        params = {
            "model": DEFAULT_MODEL,
            "messages": [{"role": "user", "content": payload.prompt}],
            "n": 1,
            "temperature": safe_temperature,
            "top_p": safe_top_p,
        }
        completion = await run_chat_completion(params)
        option_text = extract_message_content(completion)
        if option_text:
            raw_options.append(option_text)
            result = validate_options(raw_options, count)
            valid_options = result.valid
        logger.info(
            "options progress", extra={"expected": count, "received": len(valid_options), "attempts": attempts}
        )

    if len(valid_options) < count:
        logger.warning(
            "Fewer valid options generated than requested", extra={"expected": count, "received": len(valid_options)}
        )
        return JSONResponse(
            status_code=502,
            content={"error": f"Generated {len(valid_options)} of {count} options", "options": valid_options},
        )

    logger.debug("Groq options response: %s", valid_options)
    return {"options": valid_options}
